entity-service-impl/src/main/java/org/hypertrace/entity/query/service/EntityQueryServiceImpl.java

```diff
@@ -290,19 +295,128 @@ private void doUpdate(RequestContext requestContext, EntityUpdateRequest request
+      }
+    }
+  }
+
+  @Override
+  public void bulkUpdate(
+      BulkEntityUpdateRequest request, StreamObserver<ResultSetChunk> responseObserver) {
+    // Validations
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    Optional<String> tenantId = requestContext.getTenantId();
+    if (tenantId.isEmpty()) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+    if (StringUtils.isEmpty(request.getEntityType())) {
+      responseObserver.onError(new ServiceException("Entity type is missing in the request."));
+      return;
+    }
+    if (request.getEntitiesCount() == 0) {
+      responseObserver.onError(new ServiceException("Entity IDs are missing in the request."));
+    }
+
+    try {
+      doBulkUpdate(requestContext, request);
+      // Finally return the selections
+      Set<String> entityIdsList = request.getEntitiesMap().keySet();
+      List<Entity> entities =
+          getProjectedEntities(
+              entityIdsList, request.getSelectionList(), requestContext, tenantId.get());
+      responseObserver.onNext(
+          convertEntitiesToResultSetChunk(requestContext, entities, request.getSelectionList()));
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      responseObserver.onError(
+          new ServiceException("Error occurred while executing " + request, e));
+    }
+  }
+
+  private List<Entity> getProjectedEntities(
+      Iterable<String> entityIdsList,
+      List<Expression> selectionList,
+      RequestContext requestContext,
+      @NonNull String tenantId)
+      throws Exception {
+    Query entitiesQuery = Query.newBuilder().addAllEntityId(entityIdsList).build();
+    List<String> docStoreSelections =
+        entityQueryConverter.convertSelectionsToDocStoreSelections(requestContext, selectionList);
+    Iterator<Document> documentIterator =
+        entitiesCollection.search(
+            DocStoreConverter.transform(tenantId, entitiesQuery, docStoreSelections));
+    return convertDocsToEntities(documentIterator);
+  }
+
+  private void doBulkUpdate(RequestContext requestContext, BulkEntityUpdateRequest request) {
+    Map<Key, Map<String, Document>> entitiesUpdateMap = new HashMap<>();
+    Map<String, Document> documentMap = new HashMap<>();
+    for (String entityId : request.getEntities().keySet()) {
+      documentMap.clear();
+      try {
+        documentMap =
+            documentMapMaker(
+                request.getEntities().get(entityId).getUpdateOperationList(), requestContext);
+      } catch (Exception e) {
+        LOG.warn("Failed to update entity id {}", entityId);
+        continue;
+      }
+      if (!documentMap.isEmpty()) {
+        entitiesUpdateMap.put(
+            new SingleValueKey(requestContext.getTenantId().orElseThrow(), entityId), documentMap);
+      }
+    }
+
+    if (entitiesUpdateMap.isEmpty()) {
```

Comment:
shouldn't really be reachable - there should never be a successful case where we return without actually doing an update.